[UnifiedPDF] Use the scrolling tree for UnifiedPDFPlugin

hortont424 opened the pull request:
#### 34b37bd8fc5607f6eb809a7fc0ee4a5edd83a688
<pre>
[UnifiedPDF] Use the scrolling tree for UnifiedPDFPlugin
<a href="https://bugs.webkit.org/show_bug.cgi?id=265007">https://bugs.webkit.org/show_bug.cgi?id=265007</a>
<a href="https://rdar.apple.com/118154855">rdar://118154855</a>

Reviewed by Simon Fraser.

Adopt async scrolling for Unified PDF plugin, improving scrolling performance,
getting scrolling working for the first time on iOS, and paving the way towards
UI-side scrollbars on macOS.

* Source/WebCore/Sources.txt:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::scrollableAreaForScrollingNodeID const):
* Source/WebCore/page/LocalFrameView.h:
Make it possible to install a non-RenderLayer-owned ScrollableArea as associated
with a ScrollingNodeID.

* Source/WebCore/plugins/PluginViewBase.h:
(WebCore::PluginViewBase::usesAsyncScrolling):
(WebCore::PluginViewBase::scrollingNodeID):
* Source/WebCore/rendering/RenderEmbeddedObject.cpp:
(WebCore::RenderEmbeddedObject::usesAsyncScrolling):
(WebCore::RenderEmbeddedObject::scrollingNodeID):
* Source/WebCore/rendering/RenderEmbeddedObject.h:
* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::usesAsyncScrolling):
(WebKit::PluginView::scrollingNodeID):
* Source/WebKit/WebProcess/Plugins/PluginView.h:
Make it possible to ask a plugin if it wants to host a coordinated scroller,
and to retrieve the scroller&apos;s ScrollingNodeID.

* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::attachWidgetContentLayers):
Connect the plugin scrolling node as a child of the plugin hosting node.

(WebCore::RenderLayerCompositor::isLayerForPluginWithScrollCoordinatedContents const):
(WebCore::scrollCoordinationRoleForNodeType):
(WebCore::RenderLayerCompositor::detachScrollCoordinatedLayer):
(WebCore::RenderLayerCompositor::coordinatedScrollingRolesForLayer const):
(WebCore::RenderLayerCompositor::updateScrollCoordinationForLayer):
(WebCore::RenderLayerCompositor::updateScrollingNodeForPluginHostingRole):
Create a plugin hosting node if the plugin wants to host a coordinated scroller.

(WebCore::RenderLayerCompositor::updateScrollingNodeLayers):
(WebCore::RenderLayerCompositor::updateScrollingNodeForScrollingRole):
Avoid some irrelevant work which would stomp UnifiedPDFPlugin&apos;s scrolling node layers and geometry.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::wantsWheelEvents const): Deleted.
De-hoist wantsWheelEvents, since we don&apos;t want them in UnifiedPDFPlugin anymore.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::PDFPluginBase):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::UnifiedPDFPlugin):
Turn on rubber-banding.

(WebKit::UnifiedPDFPlugin::teardown):
(WebKit::UnifiedPDFPlugin::updateLayerHierarchy):
Create a plugin scrolling node, which RenderLayerCompositor will plug into the hosting node.

(WebKit::UnifiedPDFPlugin::didChangeSettings):
(WebKit::UnifiedPDFPlugin::didChangeScrollOffset):
Use boundsOrigin instead of position to scroll, matching what the UI-side scrolling code will do.

(WebKit::UnifiedPDFPlugin::updateScrollbars):
Push plugin geometry changes down to both the scrolling tree and event regions.

(WebKit::UnifiedPDFPlugin::handleWheelEvent): Deleted.

* Source/WebCore/page/scrolling/ScrollingStatePluginHostingNode.cpp: Added.
* Source/WebCore/page/scrolling/ScrollingStatePluginHostingNode.h: Added.
* Source/WebCore/page/scrolling/ScrollingStatePluginScrollingNode.cpp: Added.
* Source/WebCore/page/scrolling/ScrollingStatePluginScrollingNode.h: Added.
* Source/WebCore/page/scrolling/ScrollingTreePluginHostingNode.cpp: Added.
* Source/WebCore/page/scrolling/ScrollingTreePluginHostingNode.h: Added.
* Source/WebCore/page/scrolling/ScrollingTreePluginScrollingNode.cpp: Added.
* Source/WebCore/page/scrolling/ScrollingTreePluginScrollingNode.h: Added.
* Source/WebCore/page/scrolling/mac/ScrollingTreePluginScrollingNodeMac.h: Added.
* Source/WebCore/page/scrolling/mac/ScrollingTreePluginScrollingNodeMac.mm: Added.
* Source/WebKit/UIProcess/RemoteLayerTree/ios/ScrollingTreePluginScrollingNodeIOS.h: Added.
* Source/WebKit/UIProcess/RemoteLayerTree/ios/ScrollingTreePluginScrollingNodeIOS.mm: Added.
* Source/WebKit/UIProcess/RemoteLayerTree/mac/ScrollingTreePluginScrollingNodeRemoteMac.cpp: Added.
* Source/WebKit/UIProcess/RemoteLayerTree/mac/ScrollingTreePluginScrollingNodeRemoteMac.h: Added.
Add plugin-scrolling-specific scrolling tree nodes.

* Source/WebCore/page/scrolling/ScrollingCoordinator.h:
* Source/WebCore/page/scrolling/ScrollingCoordinatorTypes.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/page/scrolling/ScrollingCoordinatorTypes.h:
* Source/WebCore/page/scrolling/ScrollingStateNode.h:
(WebCore::ScrollingStateNode::isScrollingNode const):
(WebCore::ScrollingStateNode::isPluginScrollingNode const):
(WebCore::ScrollingStateNode::isPluginHostingNode const):
* Source/WebCore/page/scrolling/ScrollingStateTree.cpp:
(WebCore::ScrollingStateTree::createNode):
(WebCore::ScrollingStateTree::isValid const):
* Source/WebCore/page/scrolling/ScrollingTreeNode.h:
(WebCore::ScrollingTreeNode::isScrollingNode const):
(WebCore::ScrollingTreeNode::isPluginScrollingNode const):
(WebCore::ScrollingTreeNode::isPluginHostingNode const):
* Source/WebCore/page/scrolling/mac/ScrollingTreeMac.mm:
(ScrollingTreeMac::createScrollingTreeNode):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::outputPaintOrderTreeRecursive):
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::~RenderLayerBacking):
(WebCore::RenderLayerBacking::detachFromScrollingCoordinator):
(WebCore::RenderLayerBacking::setScrollingNodeIDForRole):
(WebCore::operator&lt;&lt;):
* Source/WebCore/rendering/RenderLayerBacking.h:
* Source/WebCore/rendering/RenderLayerCompositor.h:
(WebCore::allScrollCoordinationRoles):
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingCoordinatorTransaction.cpp:
(WebKit::dump):
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingCoordinatorTransaction.serialization.in:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingTree.cpp:
(WebKit::RemoteScrollingTree::createScrollingTreeNode):
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
(WebKit::RemoteScrollingCoordinatorProxyIOS::scrollViewForScrollingNodeID const):
(WebKit::RemoteScrollingCoordinatorProxyIOS::connectStateNodeLayers):
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingTreeIOS.cpp:
(WebKit::RemoteScrollingTreeIOS::createScrollingTreeNode):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.mm:
(WebKit::RemoteScrollingCoordinatorProxyMac::connectStateNodeLayers):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm:
(WebKit::RemoteScrollingTreeMac::createScrollingTreeNode):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
Plumbing for plugin-scrolling-specific scrolling tree nodes.

* Source/WebCore/rendering/EventRegion.h:
Extra exports!

Canonical link: <a href="https://commits.webkit.org/270917@main">https://commits.webkit.org/270917@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/342e24ef777d8cfe9f5280d364248b00e260157c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26811 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5427 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/28047 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/29022 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24505 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/27257 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/7261 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2818 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/24398 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/27073 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4240 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/23009 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/3721 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/3767 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/24006 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/29502 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/24451 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24407 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/30012 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/3800 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/1995 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/27920 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5252 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/23742 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6442 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/4267 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4161 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingCoordinatorTransaction.cpp:
```diff
@@ -219,6 +221,15 @@ static void dump(TextStream& ts, const ScrollingStateStickyNode& node, bool chan
 }
 
+static void dump(TextStream& ts, const ScrollingStatePluginHostingNode& node, bool changedPropertiesOnly)
+{
```
Did you mean to write?

```cpp
    dump(ts, static_cast<const ScrollingStateNode&>(node), changedPropertiesOnly);
```

hortont424 replied:
Hmm, I wonder what @smfr was on about with the `ScrollingStateFrameHostingNode` version (which this is a copy of). Will see if there's anything interesting in the output!

smfr replied:
I think there's just nothing to dump.

aprotyas commented on Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
```diff
@@ -94,6 +97,13 @@
+
+        // All ScrollingTreePluginScrollingNodes are ScrollingTreePluginScrollingNodeIOS on iOS.
+        return static_cast<ScrollingTreePluginScrollingNodeIOS*>(pluginScrollingNode)->scrollView();
+    }
```
Maybe?

```suggestion
    if (RefPtr pluginScrollingNode = dynamicDowncast<ScrollingTreePluginScrollingNode>(treeNode)) {
        // All ScrollingTreePluginScrollingNodes are ScrollingTreePluginScrollingNodeIOS on iOS.
        return static_cast<ScrollingTreePluginScrollingNodeIOS*>(pluginScrollingNode.get())->scrollView();
    }
```

hortont424 replied:
truth, this will make chris happier too

smfr commented on Source/WebCore/page/LocalFrameView.h:
```diff
@@ -950,6 +952,8 @@ class LocalFrameView final : public FrameView {
     std::unique_ptr<ListHashSet<CheckedPtr<RenderEmbeddedObject>>> m_embeddedObjectsToUpdate;
     std::unique_ptr<WeakHashSet<RenderElement>> m_slowRepaintObjects;
 
+    HashMap<ScrollingNodeID, WeakPtr<ScrollableArea>> m_scrollingNodeIDToScrollableAreaMap;
```
This deserved a comment saying "This is used for plugins", or, better, change the name so that's obvious:  `m_scrollingNodeIDToPluginScrollableAreaMap`

smfr commented on Source/WebCore/page/LocalFrameView.h:
```diff
@@ -183,6 +183,8 @@ class LocalFrameView final : public FrameView {
 
     WEBCORE_EXPORT ScrollingNodeID scrollingNodeID() const override;
     WEBCORE_EXPORT ScrollableArea* scrollableAreaForScrollingNodeID(ScrollingNodeID) const;
+    void setScrollableAreaForScrollingNodeID(ScrollingNodeID nodeID, ScrollableArea* area) { m_scrollingNodeIDToScrollableAreaMap.add(nodeID, area); }
```
`ScrollableArea&` if it's never null

smfr replied:
Should probably have "plugin" in the name.

smfr commented on Source/WebCore/page/scrolling/ScrollingStatePluginHostingNode.cpp:
```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (C) 2019-2023 Apple Inc. All rights reserved.
```
2023

hortont424 replied:
These are all right since they're copy-pastes (this was already my adjusted version :) )

smfr commented on Source/WebCore/page/scrolling/ScrollingStatePluginHostingNode.h:
```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright (C) 2019-2023 Apple Inc. All rights reserved.
```
2023

smfr commented on Source/WebCore/page/scrolling/ScrollingStatePluginScrollingNode.cpp:
```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (C) 2014-2023 Apple Inc. All rights reserved.
```
2023

smfr commented on Source/WebCore/page/scrolling/ScrollingStatePluginScrollingNode.cpp:
```diff
@@ -0,0 +1,124 @@
+namespace WebCore {
+
+ScrollingStatePluginScrollingNode::ScrollingStatePluginScrollingNode(
+    ScrollingNodeID scrollingNodeID,
```
WTAH

smfr commented on Source/WebCore/page/scrolling/ScrollingStatePluginScrollingNode.h:
```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (C) 2014-2023 Apple Inc. All rights reserved.
```
etc

smfr commented on Source/WebCore/page/scrolling/ScrollingTreePluginHostingNode.cpp:
```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (C) 2019-2023 Apple Inc. All rights reserved.
```
foo

smfr commented on Source/WebCore/rendering/RenderEmbeddedObject.h:
```diff
@@ -61,6 +61,9 @@ class RenderEmbeddedObject final : public RenderWidget {
 
     const String& pluginReplacementTextIfUnavailable() const { return m_unavailablePluginReplacementText; }
 
+    bool usesAsyncScrolling();
```
`const`

smfr commented on Source/WebCore/rendering/RenderEmbeddedObject.h:
```diff
@@ -61,6 +61,9 @@ class RenderEmbeddedObject final : public RenderWidget {
     const String& pluginReplacementTextIfUnavailable() const { return m_unavailablePluginReplacementText; }
 
+    bool usesAsyncScrolling();
+    ScrollingNodeID scrollingNodeID();
```
`const`

smfr commented on Source/WebCore/rendering/RenderLayerCompositor.cpp:
```diff
@@ -2608,6 +2608,17 @@ bool RenderLayerCompositor::attachWidgetContentLayers(RenderWidget& renderer)
+                    auto* renderEmbeddedObject = dynamicDowncast<RenderEmbeddedObject>(renderer);
+                    if (auto pluginScrollingNodeID = renderEmbeddedObject->scrollingNodeID()) {
+                        if (auto* scrollingCoordinator = this->scrollingCoordinator())
+                            scrollingCoordinator->insertNode(ScrollingNodeType::PluginScrolling, pluginScrollingNodeID, pluginHostingNodeID, 0);
```
This makes me think we might want a test that dumps the scrolling tree. We should also test dynamic changes to an `<embed>` src to and from PDF contents.

hortont424 replied:
Added this test, though cannot do "from" without just removing the <embed>

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
```diff
@@ -161,8 +162,23 @@
+                    scrollingStateNode.setScrollContainerLayer(remoteLayerTreeNode->layer());
+            }
+
+            if (scrollingStateNode.hasChangedProperty(ScrollingStateNode::Property::ScrolledContentsLayer))
```
This makes me think we need a test where the scrolled contents layer changes from non-tiled to tiled (maybe we can only do this with incremental loading).

hortont424 replied:
Do not think this is actually a thing that happens.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/ios/ScrollingTreePluginScrollingNodeIOS.h:
```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (C) 2014-2023 Apple Inc. All rights reserved.
```
Dates

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/mac/ScrollingTreePluginScrollingNodeRemoteMac.cpp:
```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (C) 2019-2023 Apple Inc. All rights reserved.
```
Dates

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/mac/ScrollingTreePluginScrollingNodeRemoteMac.cpp:
```diff
@@ -0,0 +1,71 @@
+    m_delegate->initScrollbars();
+}
+
+ScrollingTreePluginScrollingNodeRemoteMac::~ScrollingTreePluginScrollingNodeRemoteMac()
```
` = default`

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -307,14 +343,25 @@
+    if (!page)
+        return;
+    auto& scrollingCoordinator = *page->scrollingCoordinator();
+    scrollingCoordinator.setScrollingNodeScrollableAreaGeometry(m_scrollingNodeID, *this);
```
It's weird for this to happen in a function called `updateScrollbars`. I'd expect an `updateGeometry` or `sizeDidChange` function to be doing this.

rr-codes commented on Source/WebCore/page/scrolling/ScrollingStatePluginScrollingNode.cpp:
```diff
@@ -0,0 +1,124 @@
+    std::optional<WebCore::PlatformLayerIdentifier> layerID,
+    FloatSize scrollableAreaSize,
+    FloatSize totalContentsSize,
+    FloatSize reachableContentsSize,
```
what does “reachable” mean here?

hortont424 committed:
[UnifiedPDF] Use the scrolling tree for UnifiedPDFPlugin
https://bugs.webkit.org/show_bug.cgi?id=265007
rdar://118154855

Reviewed by Simon Fraser.

Adopt async scrolling for Unified PDF plugin, improving scrolling performance,
getting scrolling working for the first time on iOS, and paving the way towards
UI-side scrollbars on macOS.

* Source/WebCore/Sources.txt:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::scrollableAreaForScrollingNodeID const):
* Source/WebCore/page/LocalFrameView.h:
Make it possible to install a non-RenderLayer-owned ScrollableArea as associated
with a ScrollingNodeID.

* Source/WebCore/plugins/PluginViewBase.h:
(WebCore::PluginViewBase::usesAsyncScrolling):
(WebCore::PluginViewBase::scrollingNodeID):
* Source/WebCore/rendering/RenderEmbeddedObject.cpp:
(WebCore::RenderEmbeddedObject::usesAsyncScrolling):
(WebCore::RenderEmbeddedObject::scrollingNodeID):
* Source/WebCore/rendering/RenderEmbeddedObject.h:
* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::usesAsyncScrolling):
(WebKit::PluginView::scrollingNodeID):
* Source/WebKit/WebProcess/Plugins/PluginView.h:
Make it possible to ask a plugin if it wants to host a coordinated scroller,
and to retrieve the scroller's ScrollingNodeID.

* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::attachWidgetContentLayers):
Connect the plugin scrolling node as a child of the plugin hosting node.

(WebCore::RenderLayerCompositor::isLayerForPluginWithScrollCoordinatedContents const):
(WebCore::scrollCoordinationRoleForNodeType):
(WebCore::RenderLayerCompositor::detachScrollCoordinatedLayer):
(WebCore::RenderLayerCompositor::coordinatedScrollingRolesForLayer const):
(WebCore::RenderLayerCompositor::updateScrollCoordinationForLayer):
(WebCore::RenderLayerCompositor::updateScrollingNodeForPluginHostingRole):
Create a plugin hosting node if the plugin wants to host a coordinated scroller.

(WebCore::RenderLayerCompositor::updateScrollingNodeLayers):
(WebCore::RenderLayerCompositor::updateScrollingNodeForScrollingRole):
Avoid some irrelevant work which would stomp UnifiedPDFPlugin's scrolling node layers and geometry.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::wantsWheelEvents const): Deleted.
De-hoist wantsWheelEvents, since we don't want them in UnifiedPDFPlugin anymore.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::PDFPluginBase):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::UnifiedPDFPlugin):
Turn on rubber-banding.

(WebKit::UnifiedPDFPlugin::teardown):
(WebKit::UnifiedPDFPlugin::updateLayerHierarchy):
Create a plugin scrolling node, which RenderLayerCompositor will plug into the hosting node.

(WebKit::UnifiedPDFPlugin::didChangeSettings):
(WebKit::UnifiedPDFPlugin::didChangeScrollOffset):
Use boundsOrigin instead of position to scroll, matching what the UI-side scrolling code will do.

(WebKit::UnifiedPDFPlugin::updateScrollbars):
Push plugin geometry changes down to both the scrolling tree and event regions.

(WebKit::UnifiedPDFPlugin::handleWheelEvent): Deleted.

* Source/WebCore/page/scrolling/ScrollingStatePluginHostingNode.cpp: Added.
* Source/WebCore/page/scrolling/ScrollingStatePluginHostingNode.h: Added.
* Source/WebCore/page/scrolling/ScrollingStatePluginScrollingNode.cpp: Added.
* Source/WebCore/page/scrolling/ScrollingStatePluginScrollingNode.h: Added.
* Source/WebCore/page/scrolling/ScrollingTreePluginHostingNode.cpp: Added.
* Source/WebCore/page/scrolling/ScrollingTreePluginHostingNode.h: Added.
* Source/WebCore/page/scrolling/ScrollingTreePluginScrollingNode.cpp: Added.
* Source/WebCore/page/scrolling/ScrollingTreePluginScrollingNode.h: Added.
* Source/WebCore/page/scrolling/mac/ScrollingTreePluginScrollingNodeMac.h: Added.
* Source/WebCore/page/scrolling/mac/ScrollingTreePluginScrollingNodeMac.mm: Added.
* Source/WebKit/UIProcess/RemoteLayerTree/ios/ScrollingTreePluginScrollingNodeIOS.h: Added.
* Source/WebKit/UIProcess/RemoteLayerTree/ios/ScrollingTreePluginScrollingNodeIOS.mm: Added.
* Source/WebKit/UIProcess/RemoteLayerTree/mac/ScrollingTreePluginScrollingNodeRemoteMac.cpp: Added.
* Source/WebKit/UIProcess/RemoteLayerTree/mac/ScrollingTreePluginScrollingNodeRemoteMac.h: Added.
Add plugin-scrolling-specific scrolling tree nodes.

* Source/WebCore/page/scrolling/ScrollingCoordinator.h:
* Source/WebCore/page/scrolling/ScrollingCoordinatorTypes.cpp:
(WebCore::operator<<):
* Source/WebCore/page/scrolling/ScrollingCoordinatorTypes.h:
* Source/WebCore/page/scrolling/ScrollingStateNode.h:
(WebCore::ScrollingStateNode::isScrollingNode const):
(WebCore::ScrollingStateNode::isPluginScrollingNode const):
(WebCore::ScrollingStateNode::isPluginHostingNode const):
* Source/WebCore/page/scrolling/ScrollingStateTree.cpp:
(WebCore::ScrollingStateTree::createNode):
(WebCore::ScrollingStateTree::isValid const):
* Source/WebCore/page/scrolling/ScrollingTreeNode.h:
(WebCore::ScrollingTreeNode::isScrollingNode const):
(WebCore::ScrollingTreeNode::isPluginScrollingNode const):
(WebCore::ScrollingTreeNode::isPluginHostingNode const):
* Source/WebCore/page/scrolling/mac/ScrollingTreeMac.mm:
(ScrollingTreeMac::createScrollingTreeNode):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::outputPaintOrderTreeRecursive):
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::~RenderLayerBacking):
(WebCore::RenderLayerBacking::detachFromScrollingCoordinator):
(WebCore::RenderLayerBacking::setScrollingNodeIDForRole):
(WebCore::operator<<):
* Source/WebCore/rendering/RenderLayerBacking.h:
* Source/WebCore/rendering/RenderLayerCompositor.h:
(WebCore::allScrollCoordinationRoles):
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingCoordinatorTransaction.cpp:
(WebKit::dump):
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingCoordinatorTransaction.serialization.in:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingTree.cpp:
(WebKit::RemoteScrollingTree::createScrollingTreeNode):
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
(WebKit::RemoteScrollingCoordinatorProxyIOS::scrollViewForScrollingNodeID const):
(WebKit::RemoteScrollingCoordinatorProxyIOS::connectStateNodeLayers):
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingTreeIOS.cpp:
(WebKit::RemoteScrollingTreeIOS::createScrollingTreeNode):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.mm:
(WebKit::RemoteScrollingCoordinatorProxyMac::connectStateNodeLayers):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm:
(WebKit::RemoteScrollingTreeMac::createScrollingTreeNode):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
Plumbing for plugin-scrolling-specific scrolling tree nodes.

* Source/WebCore/rendering/EventRegion.h:
Extra exports!

Canonical link: https://commits.webkit.org/270917@main
